client/question/services/ConversationManagerService.js

```diff
@@ -13,18 +13,48 @@
 // limitations under the License.
 
 /**
- * @fileoverview Service that orchestrates the processing of a user's code
- * submission.
+ * @fileoverview Service that orchestrates TIE's conversation with the learner.
  */
 
-tie.factory('SolutionHandlerService', [
+tie.factory('ConversationManagerService', [
   '$q', 'CodePreprocessorDispatcherService', 'CodeRunnerDispatcherService',
   'FeedbackGeneratorService', 'PrereqCheckDispatcherService',
-  'TranscriptService', 'CodeSubmissionObjectFactory',
+  'TranscriptService', 'CodeSubmissionObjectFactory', 'FEEDBACK_CATEGORIES',
+  'FeedbackDetailsObjectFactory',
   function(
       $q, CodePreprocessorDispatcherService, CodeRunnerDispatcherService,
       FeedbackGeneratorService, PrereqCheckDispatcherService,
-      TranscriptService, CodeSubmissionObjectFactory) {
+      TranscriptService, CodeSubmissionObjectFactory, FEEDBACK_CATEGORIES,
+      FeedbackDetailsObjectFactory) {
+
+    // TODO(sll): Modify this to also depend on the learner's submission
+    // history.
+
+    /**
+     * Determines the feedback details that fully characterize the feedback
+     * that the learner should receive.
+     *
+     * @returns {FeedbackDetails}
+     */
+    var computeFeedbackDetails = function(codeEvalResult) {
```

Comment:
So I'm going to end up implementing a similar version of this for the compile step, since we'll need to make sure we have similar checks for the compile step (except it would return a SyntaxError rather than a RuntimeError).

Not to put my work on you, but is there a way you could generalize this such that I can use the same method, or at least call it something that makes it obvious that it's FeedbackDetails from running a user's code, rather than compiling it?

Reply:
Done -- added an "enum" to pass in information about the context. Thanks for catching this!